vms/platformvm/service.go

```diff
@@ -704,56 +704,118 @@ func (s *Service) loadStakerTxAttributes(txID ids.ID) (*stakerAttributes, error)
 // GetCurrentValidators returns the current validators. If a single nodeID
 // is provided, full delegators information is also returned. Otherwise only
 // delegators' number and total weight is returned.
-func (s *Service) GetCurrentValidators(_ *http.Request, args *GetCurrentValidatorsArgs, reply *GetCurrentValidatorsReply) error {
+func (s *Service) GetCurrentValidators(r *http.Request, args *GetCurrentValidatorsArgs, reply *GetCurrentValidatorsReply) error {
 	s.vm.ctx.Log.Debug("API called",
 		zap.String("service", "platform"),
 		zap.String("method", "getCurrentValidators"),
 	)
 
-	reply.Validators = []interface{}{}
-
-	// Validator's node ID as string --> Delegators to them
-	vdrToDelegators := map[ids.NodeID][]platformapi.PrimaryDelegator{}
-
 	// Create set of nodeIDs
 	nodeIDs := set.Of(args.NodeIDs...)
 
 	s.vm.ctx.Lock.Lock()
 	defer s.vm.ctx.Lock.Unlock()
 
+	// Check if subnet is L1
+	_, err := s.vm.state.GetSubnetToL1Conversion(args.SubnetID)
+	if err == database.ErrNotFound {
+		// Subnet is not L1, get validators for the subnet
+		reply.Validators, err = s.getPrimaryOrSubnetValidators(
+			args.SubnetID,
+			nodeIDs,
+		)
+		return err
+	}
+	if err != nil {
+		return err
+	}
+
+	// Subnet is L1, get validators for L1
+	reply.Validators, err = s.getL1Validators(
+		r.Context(),
+		args.SubnetID,
+		nodeIDs,
+	)
+	return err
+}
+
+func (s *Service) getL1Validators(
+	ctx context.Context,
+	subnetID ids.ID,
+	nodeIDs set.Set[ids.NodeID],
+) ([]interface{}, error) {
+	validators := []interface{}{}
+	baseStakers, l1Validators, _, err := s.vm.state.GetCurrentValidators(ctx, subnetID)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, staker := range baseStakers {
+		if nodeIDs.Len() != 0 && !nodeIDs.Contains(staker.NodeID) {
+			continue
+		}
+
+		apiStaker := toPlatformStaker(staker)
+		validators = append(validators, apiStaker)
+	}
+
+	for _, l1Validator := range l1Validators {
+		if nodeIDs.Len() != 0 && !nodeIDs.Contains(l1Validator.NodeID) {
```

Comment:
general thought, not related to this PR's purpose - this filter where we expect to fetch all validators if the nodeID is empty cannot be inferred unless you scroll up to the argument description. 

Might be worthwhile to explicitly create a variable that coveys this intent instead of the reader of the code needing domain knowledge about the expected behavior.

Reply:
https://github.com/ava-labs/avalanchego/pull/3671